### DISCUSSION TITLE
How to convert a `System.Drawing.Bitmap` to `Microsoft.UI.Xaml.Media.ImageSource`?

### DISCUSSION THREAD

**Discussion by kyoyama-kazusa** [microsoft/WindowsAppSDK#1374] [Q&A] [answered]
Hello. I'm the new for using WinUI 3 on .NET 6.

I've implemented a data structure that automatically generates a picture that is of type `System.Drawing.Bitmap` using GDI+, but I found that a `Micorost.UI.Xaml.Controls.Image` control must receive a value `ImageSource`.

I tried to search for solutions on Stackoverflow. Although I found a possible solution to convert, that solution will introduce a new type `Int32Rect`, which doesn't exist in my solution and I can't import or download from Nuget packages.

So... I don't know how to convert a `Bitmap` to `ImageSource`. Is there any ideas to convert?

Thanks.

**Comment by huoyaoyuan** [comment, marked as answer]
There should be no way other than building from pixels. The suggested solution is not to use GDI+ and `System.Drawing` at all.
It could be even easier to save the image into some file format (.bmp/.png).

**Reply by kyoyama-kazusa** [reply]
Thanks for your reply.

To be honest... I'm working on sudokus (a puzzle game, "number place" as its alias), about solving or generating sudokus. The program will update the status of the displaying of a sudoku **very frequently**. I found that GDI+ is a simple and convenient way to implement on this. If I use GDI+, the procedure will create a file firstly, and display it then. In this way the algorithm may effect the performance. 😃 

It seems that I have to use WPF `Shape` type instance to implement on this.

**Reply by huoyaoyuan** [reply]
I'm very familiar with sudoku.

The tiles are updated **very infrequently** - they won't move at all. So any update will only change the tile itself. Sudoku is a data intensive game and suitable for MVVM.

**Reply by kyoyama-kazusa** [reply]
Thanks. I'll try. 😺
